govuk-prototype-kit.config.json

```
@@ -2,14 +2,12 @@
   "assets": "/lib/assets/images",
   "nunjucksPaths": "/lib/nunjucks",
   "nunjucksFilters": "/lib/filters/coreFilters",
+  "sass": "/lib/assets/sass/internal/manage-prototype.scss",
   "scripts": [
     "/lib/assets/javascripts/kit.js",
     "/lib/assets/javascripts/auto-store-data.js"
   ],
-  "sass": [
-    "/lib/assets/sass/internal/manage-prototype.scss"
-  ],
-  "autoImportNunjucksComponents": [
+  "importNunjucksMacrosInto": [
```

I have some thoughts about this name... I think if we wanted in future to do things to layout files more than just add Nunjucks import lines, would we need an extra key that would probably be redundant?

I'd suggest `layouts` instead, ran this past @joelanman as well.

```suggestion
  "layouts": [
```

---

I don't think `layouts` is the right name for two reasons:

1. I don't think every layout will have the nunjucks includes added to it
2. There might be a time where we allow plugins to specify layouts in the same way as they can specify templates - giving the kit user a list of layouts they can choose from, this would be a different interface requiring a name etc.  If we do that later it would make sense to put importMacros or something similar in there but I don't think it would be helpful to assume that everything that has the macros auto-loaded should be in that list

I'm concerned about `layouts` for its ambiguity as I think it will cause us problems when we include more options in the configuration.  It's much easier to phase something specific out than to work around something that was too general.

To describe what this does:  When a file is in this list a copy of it is made with the Nunjucks macros in, this copy will be available to the user under the same name as the original file but with a higher priority (it will be chosen instead of the template provided).  The old approach was to [include these in `layout.html` so that anything that extends it picks up the includes](https://github.com/alphagov/govuk-prototype-kit/blob/9e4a7fd6a4689be14c067c674bdf0f767fd03f1f/app/views/layout.html#L6-L33).

I'll continue to advocate for a specific name describing what we'll do with the files in the list rather than assuming that everything in the list will be a layout and that every template will need this.

---

> I'm concerned about layouts for its ambiguity as I think it will cause us problems when we include more options in the configuration. It's much easier to phase something specific out than to work around something that was too general.

I'm not sure I'd agree with this, I think it would be easier to add config options to the `layouts` list, similar to the way we have the `templates` list. For example, to have a layout without including the Nunjucks imports:

```
layouts = [
  {
    "path": "/lib/nunjucks/govuk-prototype-kit/layouts/govuk-branded.html",
  },
  {
    "path": "/lib/nunjucks/govuk-prototype-kit/layouts/unbranded.html",
    "autoImportNunjucksComponents": false
  }
]
```

We can maintain backwards compatibility by allowing the list to be either strings or obiects with configuration parameters. Having a configuration option which has a type similar to `Array<string | Object>` is a fairly common in JavaScript I think? As you say this leans more towards making it easier to add new features without plugin authors having to update their plugins, but I think that is preferable, especially if we want to make smaller changes than adding a layouts page.

My other concern about the name `importNunjucksMacrosInto` is that it took me 3 minutes trying to figure out what it meant 😅 Which might be just because I was jet-lagged, but I think is related to the fact that the name is kind of backwards to the context of it being a Nunjucks template.

---

Perhaps an alternative approach would be to have a generated template that layouts can include. This would give layout files complete control over whether and where the macros are imported. For instance:

```
# layout.html

{% extends "govuk/template.njk" %}

{% include "govuk-prototype-kit/includes/allMacros.njk" %}

...
```

I think in my mind it is much clearer what the outcome of this is.

---

My first approach was to try `{% include "govuk-prototype-kit/includes/allMacros.njk" %}` but the imports don't carry through so that's not an option.

I agree with the way that the `layouts` could grow but I think we're mixing concepts, I don't believe there's a 1:1 match between "Nunjucks files which can be used as layouts" and "Nunjucks files to include all macros into" - I think they're two distinct but heavily overlapping groups.  If we do develop a clear understanding about what we do with `layouts` later then we can include a configuration option to import all the macros.

In terms of future-proofing the API I really want to avoid `layouts` because I can see us creating a meaning for that later.  I'd like this to be a more specific name as it's much easier to deprecate a specific option than support two different uses of the same configuration option - to be clear, I'm really talking here about the things that aren't layouts but want to include all the macros.  I'd like to try to find a specific name for this feature which could in the future co-exist with `layouts`, I'm assuming we'd have a list of layouts shown to the user like we have a list of templates and I wouldn't want to have to figure out whether things in `layouts` are really layouts or if they're just there to import the macros.

When considering the name I think it needs:

 - To have the word Nunjucks
 - To have the word Macros
 - To have the word Include

Therefore the shortest nae that meets that criteria would be `includNunjucksMacros`.

Shall we discuss the criteria I've set for the name and then discuss the name itself?  This configuration file might exist in a kit which supports multiple templating languages (e.g. Markdown) as well as plugins providing macros that aren't "components" by the GOV.UK Frontend definition.

---

I dont have a strong opinion here - I just thought `includeNunjucksMacros` read like a boolean. Another suggestion - given we are the only ones currently using this property, do we need to expose it yet? We could do it internally and give ourselves more time to consider the use cases

---

> 
My first approach was to try {% include "govuk-prototype-kit/includes/allMacros.njk" %} but the imports don't carry through so that's not an option.

Ah. My bad, I should have checked this would work before suggesting it.

> I agree with the way that the layouts could grow but I think we're mixing concepts, I don't believe there's a 1:1 match between "Nunjucks files which can be used as layouts" and "Nunjucks files to include all macros into" - I think they're two distinct but heavily overlapping groups.

I think this is where we're disagreeing fundamentally, but you've obviously spent more time thinking about it than me, so I will defer to you on this. The last thing I will ask though, what are the circumstance we can think of where users wouldn't want all macros available? How likely are these circumstances?

> When considering the name I think it needs:
>
> - To have the word Nunjucks
> - To have the word Macros
> - To have the word Include

I'm not totally sold on the word 'include', but the other two criteria make sense to me.

At this point I think we're getting into bikeshedding though (my fault, I realise). If it the option has to be in the configuration file, and we don't want it to be an option within a list of objects, and we don't care for it to stick around too long, then I suppose any name will do, as long as it is documented.

---

> What are the circumstance we can think of where users wouldn't want all macros available? How likely are these circumstances?

I've gone through two separate thought processes on this this morning.  One is to think about how it affects the solution if we decide this case doesn't need covering, the other is to think about how this might cause a problem.

My thinking before is that plugin developers might not need the macros on all layouts - not needing them but them being present doesn't really cause a problem but I did find one case where them being there can be annoying (I'll put that at the end of this comment).  So the big reason why I want this to be opt-in is because it's us re-writing someone's Nunjucks file and referencing our rewritten version, it feels too intrusive to do without the plugin developer asking for it.

The other thought was about how it would change the interface if this was wanted on every layout, I'll still suggest a name which isn't `layout` because of the [potential for] cases which aren't layouts and the number of unknowns around what we'll want to do with layouts in the future.

> At this point I think we're getting into bikeshedding though (my fault, I realise)

It's been really useful to go through this.

> If it the option has to be in the configuration file, and we don't want it to be an option within a list of objects, and we don't care for it to stick around too long, then I suppose any name will do, as long as it is documented.

This is a pretty good reflection of how I'm seeing this - it's not perfect but it won't get in the way later.  I don't think there would be anything wrong with it being a list of objects but I think it should be a really specific name for this feature to keep our options open further down the line - if it's a specific key then it's less likely to benefit from being an object.

---

The example.  If I build a plugin which has a view ready to go, in the current `v13` kit this would be used via a router and `res.render('natalies-wonderful-plugin/views/some-useful-page-you-might-want-to-use.html')`.  When I'm building that I don't want to reference any components from other plugins, everything in my plugin is self-contained.  I have a layout and this view that references the layout:

```
{% extends "natalies-wonderful-plugin/layouts/default.html" %}

{% from "natalies-wonderful-components/someComponent.njk" import someComponent %}

{% block content %}
  {{ someComponent({ option: 1} ) }}
{% endblock %}

{% block footer %}
  {{ someComponent({ option: 2 }) }}
{% endblock %}
```

The user installs another plugin which contains a nunjucks macro called `someComponent`, now their version of `someComponent` appears instead of mine.  To resolve this I need to update my view (within the plugin) to be:

```
{% extends "natalies-wonderful-plugin/layouts/default.html" %}

{% block content %}
  {% from "natalies-wonderful-components/someComponent.njk" import someComponent %}
  {{ someComponent({ option: 1} ) }}
{% endblock %}

{% block footer %}
  {% from "natalies-wonderful-components/someComponent.njk" import someComponent %}
  {{ someComponent({ option: 2 }) }}
{% endblock %}
```

Because if my import is outside a block it takes a lower priority than the imports in the layout file.  In this scenario I don't want the macros from other plugins so I want to turn off that option.

With the PR as it stands this won't come up for the plugin developer because they wouldn't actively activate the importing of all the macros - they wouldn't choose to opt in to a feature they're not using.  In a future version they might list their layout in our `layouts` and the macros might be included by default but they should be able to opt-out of that if they don't want to risk this sort of behaviour.

FYI - my test code for this was to create two files in the kit (if you want to reproduce it yourself):

1. `index.html`

```
{% extends "govuk-prototype-kit/layouts/govuk-branded.html" %}

{% from "myCharCount.njk" import govukCharacterCount %}

{% block content %}
  {{ govukCharacterCount() }}
{% endblock %}
```

2. `myCharCount.njk`

```
{% macro govukCharacterCount(params) %}
  <h2>Hi mum</h2>
{% endmacro %}
```

To be clear - I'm not saying this is likely, I'm just saying that there shouldn't be an assumption that everyone will benefit from having the components imported into layouts.
---

---

Okay, I'm considering this thread resolved and approving the PR.